crates/db_connection_pool/src/dbconnection/snowflakeconn.rs

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::datatypes::{Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::physical_plan::memory::MemoryStream;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::sql::TableReference;
+use futures::stream;
+use snafu::prelude::*;
+use snowflake_api::SnowflakeApi;
+
+use super::AsyncDbConnection;
+use super::DbConnection;
+use super::Result;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Not implemented"))]
+    NotImplemented,
+
+    #[snafu(display("Unable to retrieve schema"))]
+    UnableToRetrieveSchema,
+
+    #[snafu(display("Unexpected query response, expected Arrow, got JSON"))]
+    UnexpectedResponse,
+
+    #[snafu(display("Error executing query: {source}"))]
+    SnowflakeQueryError {
+        source: snowflake_api::SnowflakeApiError,
+    },
+}
+
+pub struct SnowflakeConnection {
+    pub api: Arc<SnowflakeApi>,
+}
+
+impl<'a> DbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConnection {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn as_async(&self) -> Option<&dyn super::AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)>> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl<'a> AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConnection {
+    fn new(api: Arc<SnowflakeApi>) -> Self {
+        SnowflakeConnection { api }
+    }
+
+    async fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef> {
+        let table = table_reference.to_quoted_string();
+
+        let res = self
+            .api
+            .exec(format!("SELECT * FROM {table} limit 1").as_str())
+            .await
+            .context(SnowflakeQuerySnafu)?;
+
+        match res {
+            snowflake_api::QueryResult::Arrow(record_batches) => {
+                let schema = record_batches[0].schema();
+                return Ok(Arc::clone(&schema));
+            }
+            _ => UnableToRetrieveSchemaSnafu.fail()?,
```

Comment:
do we miss any information here why it fails?

Reply:
Updated to include more details. Thank you!

```rust
match res {
      snowflake_api::QueryResult::Arrow(record_batches) => {
          let schema = record_batches[0].schema();
          return Ok(Arc::clone(&schema));
      }
      snowflake_api::QueryResult::Empty => UnableToRetrieveSchemaSnafu {
          reason: "query to get table schema returned no result".to_string(),
      }
      .fail()?,
      snowflake_api::QueryResult::Json(json) => UnexpectedResponseSnafu {
          json: json.to_string(),
      }
      .fail()?,
  }
```